zebra-chain/src/notes/sprout/nullifiers.rs

```diff
@@ -0,0 +1,100 @@
+#![allow(dead_code)]
+
+use std::io;
+
+use byteorder::{ByteOrder, LittleEndian};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    keys::sprout::SpendingKey,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+};
+
+/// PRF^nf is used to derive a Sprout nullifer from the receiver's
+/// spending key a_sk and a nullifier seed ρ, instantiated using the
+/// SHA-256 compression function.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#abstractprfs
+/// https://zips.z.cash/protocol/protocol.pdf#commitmentsandnullifiers
+fn prf_nf(a_sk: [u8; 32], rho: [u8; 32]) -> [u8; 32] {
+    let mut state = [0u32; 8];
+    let mut block = [0u8; 64];
+
+    block[0..32].copy_from_slice(&a_sk[..]);
+    // The first four bits –i.e. the most signicant four bits of the
+    // first byte– are used to separate distinct uses
+    // of SHA256Compress, ensuring that the functions are independent.
+    block[0] |= 0b1110_0000;
```

Comment:
Since you're representing a_sk as `[u8; 32]`, this might do the wrong thing if the first byte of a_sk wasn't correctly masked. The zcashd code has a uint252 type that ensures correct masking.

Reply:
Hmm, do we have a sprout spending key type? could this code use that?

Reply:
See comment below: https://github.com/ZcashFoundation/zebra/pull/596/files#r467740903

Reply:
I'm keeping this as-is for now, but I've created #900 to look at this independently.